File: internal/model/log.go
package model

type AttemptStatus string

const (
	AttemptSuccess      AttemptStatus = "success"
	AttemptFailed       AttemptStatus = "failed"
	AttemptCircuitBreak AttemptStatus = "circuit_break"
	AttemptSkipped      AttemptStatus = "skipped"
)

type ChannelAttempt struct {
	ChannelID        int           `json:"channel_id"`
	ChannelKeyID     int           `json:"channel_key_id,omitempty"`
	KeyID            int           `json:"key_id,omitempty"`
	ChannelName      string        `json:"channel_name"`
	SiteID           int           `json:"site_id,omitempty"`
	SiteAccountID    int           `json:"site_account_id,omitempty"`
	AccountID        int           `json:"account_id,omitempty"`
	BaseURL          string        `json:"base_url,omitempty"`
	ModelName        string        `json:"model_name"`
	UpstreamModel    string        `json:"upstream_model,omitempty"`
	RequestProtocol  string        `json:"request_protocol,omitempty"`
	UpstreamProtocol string        `json:"upstream_protocol,omitempty"`
	ResponseProtocol string        `json:"response_protocol,omitempty"`
	AttemptNum       int           `json:"attempt_num"`
	AttemptIndex     int           `json:"attempt_index,omitempty"`
	Status           AttemptStatus `json:"status"`
	Duration         int           `json:"duration"`
	DurationMS       int           `json:"duration_ms,omitempty"`
	TTFBMS           int           `json:"ttfb_ms,omitempty"`
	TotalMS          int           `json:"total_ms,omitempty"`
	HTTPStatus       int           `json:"http_status,omitempty"`
	FailureReason    string        `json:"failure_reason,omitempty"`
	Retryable        bool          `json:"retryable,omitempty"`
	InputTokens      int           `json:"input_tokens,omitempty"`
	OutputTokens     int           `json:"output_tokens,omitempty"`
	CacheTokens      int           `json:"cache_tokens,omitempty"`
	InputCost        float64       `json:"input_cost,omitempty"`
	OutputCost       float64       `json:"output_cost,omitempty"`
	EstimatedCost    float64       `json:"estimated_cost,omitempty"`
	CostIncurred     string        `json:"cost_incurred,omitempty"`
	CostSource       string        `json:"cost_source,omitempty"`
	ServiceTier      string        `json:"service_tier,omitempty"`
	ErrorSummary     string        `json:"error_summary,omitempty"`
	CreatedAt        int64         `json:"created_at,omitempty"`
	Sticky           bool          `json:"sticky,omitempty"`
	Msg              string        `json:"msg,omitempty"`
}

type RelayLogWSMode string

const (
	RelayLogWSModeFresh        RelayLogWSMode = "fresh"
	RelayLogWSModeContinuation RelayLogWSMode = "continuation"
	RelayLogWSModeReplay       RelayLogWSMode = "replay"
)

type RelayLogWSRecovery string

const (
	RelayLogWSRecoveryReconnect RelayLogWSRecovery = "reconnect"
	RelayLogWSRecoveryReplay    RelayLogWSRecovery = "replay"
	RelayLogWSRecoveryDowngrade RelayLogWSRecovery = "downgrade"
)

type RelayLog struct {
	ID                   int64               `json:"id" gorm:"primaryKey;autoIncrement:false"`
	TraceID              string              `json:"trace_id" gorm:"index"`
	ThreadID             string              `json:"thread_id,omitempty"`
	ClientAPIKeyID       int                 `json:"client_api_key_id,omitempty" gorm:"index"`
	GroupID              int                 `json:"group_id,omitempty" gorm:"index"`
	Time                 int64               `json:"time"`
	RequestModelName     string              `json:"request_model_name"`
	RequestAPIKeyName    string              `json:"request_api_key_name"`
	RequestStream        bool                `json:"request_stream,omitempty" gorm:"index"`
	RequestSource        string              `json:"request_source,omitempty" gorm:"index"`
	ClientIP             string              `json:"client_ip,omitempty" gorm:"index"`
	ChannelId            int                 `json:"channel" gorm:"index"`
	ChannelName          string              `json:"channel_name"`
	ActualModelName      string              `json:"actual_model_name"`
	FinalStatus          string              `json:"final_status,omitempty" gorm:"index"`
	FinalChannelID       int                 `json:"final_channel_id,omitempty"`
	FinalSiteID          int                 `json:"final_site_id,omitempty"`
	FinalUpstreamModel   string              `json:"final_upstream_model,omitempty"`
	AttemptsCount        int                 `json:"attempts_count,omitempty"`
	TotalLatencyMS       int                 `json:"total_latency_ms,omitempty"`
	InputTokens          int                 `json:"input_tokens"`
	TransportInputTokens *int                `json:"transport_input_tokens,omitempty"`
	BillInputTokens      *int                `json:"bill_input_tokens,omitempty"`
	CacheReadTokens      *int                `json:"cache_read_tokens,omitempty"`
	CacheWriteTokens     *int                `json:"cache_write_tokens,omitempty"`
	CacheTokens          int                 `json:"cache_tokens,omitempty"`
	OutputTokens         int                 `json:"output_tokens"`
	Ftut                 int                 `json:"ftut"`
	UseTime              int                 `json:"use_time"`
	Cost                 float64             `json:"cost"`
	EstimatedCost        float64             `json:"estimated_cost,omitempty"`
	FinalSuccessCost     float64             `json:"final_success_cost,omitempty"`
	TotalAttemptCost     float64             `json:"total_attempt_cost,omitempty"`
	FailedAttemptCost    float64             `json:"failed_attempt_estimated_cost,omitempty"`
	ServiceTier          string              `json:"service_tier,omitempty"`
	RequestContent       string              `json:"request_content"`
	ResponseContent      string              `json:"response_content"`
	Error                string              `json:"error"`
	Attempts             []ChannelAttempt    `json:"attempts" gorm:"serializer:json"`
	TotalAttempts        int                 `json:"total_attempts"`
	UsedWS               bool                `json:"used_ws" gorm:"default:false"`
	WSMode               *RelayLogWSMode     `json:"ws_mode,omitempty"`
	WSRecovery           *RelayLogWSRecovery `json:"ws_recovery,omitempty"`
}

type RequestTrace struct {
	ID                 int64   `json:"id" gorm:"primaryKey;autoIncrement:false"`
	TraceID            string  `json:"trace_id" gorm:"index"`
	RelayLogID         int64   `json:"relay_log_id" gorm:"uniqueIndex"`
	ThreadID           string  `json:"thread_id,omitempty" gorm:"index"`
	ClientAPIKeyID     int     `json:"client_api_key_id,omitempty" gorm:"index"`
	GroupID            int     `json:"group_id,omitempty" gorm:"index"`
	ClientModel        string  `json:"client_model" gorm:"index"`
	RequestSource      string  `json:"request_source,omitempty" gorm:"index"`
	RequestStream      bool    `json:"request_stream,omitempty" gorm:"index"`
	ClientIP           string  `json:"client_ip,omitempty" gorm:"index"`
	FinalStatus        string  `json:"final_status,omitempty" gorm:"index"`
	FinalChannelID     int     `json:"final_channel_id,omitempty" gorm:"index"`
	FinalSiteID        int     `json:"final_site_id,omitempty" gorm:"index"`
	FinalUpstreamModel string  `json:"final_upstream_model,omitempty" gorm:"index"`
	AttemptsCount      int     `json:"attempts_count,omitempty"`
	TotalLatencyMS     int     `json:"total_latency_ms,omitempty"`
	InputTokens        int     `json:"input_tokens,omitempty"`
	OutputTokens       int     `json:"output_tokens,omitempty"`
	CacheTokens        int     `json:"cache_tokens,omitempty"`
	EstimatedCost      float64 `json:"estimated_cost,omitempty"`
	FinalSuccessCost   float64 `json:"final_success_cost,omitempty"`
	TotalAttemptCost   float64 `json:"total_attempt_cost,omitempty"`
	FailedAttemptCost  float64 `json:"failed_attempt_estimated_cost,omitempty"`
	ServiceTier        string  `json:"service_tier,omitempty"`
	UsedWS             bool    `json:"used_ws,omitempty" gorm:"index"`
	WSMode             string  `json:"ws_mode,omitempty"`
	WSRecovery         string  `json:"ws_recovery,omitempty"`
	CreatedAt          int64   `json:"created_at" gorm:"index"`
}

type RequestAttempt struct {
	ID               int64         `json:"id" gorm:"primaryKey;autoIncrement:false"`
	TraceID          string        `json:"trace_id" gorm:"index"`
	RelayLogID       int64         `json:"relay_log_id" gorm:"index"`
	AttemptIndex     int           `json:"attempt_index,omitempty" gorm:"index"`
	AttemptNum       int           `json:"attempt_num,omitempty"`
	ChannelID        int           `json:"channel_id,omitempty" gorm:"index"`
	ChannelKeyID     int           `json:"channel_key_id,omitempty" gorm:"index"`
	KeyID            int           `json:"key_id,omitempty" gorm:"index"`
	ChannelName      string        `json:"channel_name,omitempty"`
	SiteID           int           `json:"site_id,omitempty" gorm:"index"`
	SiteAccountID    int           `json:"site_account_id,omitempty" gorm:"index"`
	AccountID        int           `json:"account_id,omitempty" gorm:"index"`
	BaseURL          string        `json:"base_url,omitempty"`
	ModelName        string        `json:"model_name,omitempty" gorm:"index"`
	UpstreamModel    string        `json:"upstream_model,omitempty" gorm:"index"`
	RequestProtocol  string        `json:"request_protocol,omitempty" gorm:"index"`
	UpstreamProtocol string        `json:"upstream_protocol,omitempty" gorm:"index"`
	ResponseProtocol string        `json:"response_protocol,omitempty" gorm:"index"`
	Status           AttemptStatus `json:"status,omitempty" gorm:"index"`
	HTTPStatus       int           `json:"http_status,omitempty" gorm:"index"`
	FailureReason    string        `json:"failure_reason,omitempty" gorm:"index"`
	Retryable        bool          `json:"retryable,omitempty" gorm:"index"`
	DurationMS       int           `json:"duration_ms,omitempty"`
	TTFBMS           int           `json:"ttfb_ms,omitempty"`
	TotalMS          int           `json:"total_ms,omitempty"`
	InputTokens      int           `json:"input_tokens,omitempty"`
	OutputTokens     int           `json:"output_tokens,omitempty"`
	CacheTokens      int           `json:"cache_tokens,omitempty"`
	InputCost        float64       `json:"input_cost,omitempty"`
	OutputCost       float64       `json:"output_cost,omitempty"`
	EstimatedCost    float64       `json:"estimated_cost,omitempty"`
	CostIncurred     string        `json:"cost_incurred,omitempty"`
	CostSource       string        `json:"cost_source,omitempty"`
	ServiceTier      string        `json:"service_tier,omitempty"`
	ErrorSummary     string        `json:"error_summary,omitempty"`
	Sticky           bool          `json:"sticky,omitempty"`
	CreatedAt        int64         `json:"created_at" gorm:"index"`
}

type RelayLogListQuery struct {
	Page          int
	PageSize      int
	StartTime     *int
	EndTime       *int
	TimeRange     string
	ChannelIDs    []int
	Model         string
	TraceID       string
	APIKey        string
	APIKeyID      *int
	Status        string
	HTTPStatus    string
	FailureReason string
	Protocol      string
	Source        string
	Stream        *bool
	Failover      *bool
	CacheHit      *bool
	SortBy        string
	SortOrder     string
	IncludeBody   bool
}

type RelayLogListResult struct {
	Items    []RelayLog `json:"items"`
	Total    int        `json:"total"`
	Page     int        `json:"page"`
	PageSize int        `json:"page_size"`
	HasMore  bool       `json:"has_more"`
}

type RequestTraceListQuery struct {
	Page          int
	PageSize      int
	StartTime     *int
	EndTime       *int
	TimeRange     string
	ChannelIDs    []int
	Model         string
	TraceID       string
	APIKeyID      *int
	Status        string
	HTTPStatus    string
	FailureReason string
	Protocol      string
	Source        string
	Stream        *bool
	Failover      *bool
	SortBy        string
	SortOrder     string
}

type RequestTraceListResult struct {
	Items    []RequestTrace `json:"items"`
	Total    int64          `json:"total"`
	Page     int            `json:"page"`
	PageSize int            `json:"page_size"`
	HasMore  bool           `json:"has_more"`
}

type RequestTraceAuditBucket struct {
	Key          string  `json:"key" gorm:"column:key"`
	Count        int64   `json:"count" gorm:"column:count"`
	InputTokens  int64   `json:"input_tokens" gorm:"column:input_tokens"`
	OutputTokens int64   `json:"output_tokens" gorm:"column:output_tokens"`
	CacheTokens  int64   `json:"cache_tokens" gorm:"column:cache_tokens"`
	TotalTokens  int64   `json:"total_tokens" gorm:"column:total_tokens"`
	Cost         float64 `json:"cost" gorm:"column:cost"`
	AvgLatency   float64 `json:"avg_latency_ms" gorm:"column:avg_latency_ms"`
}

type RequestTraceAuditSummary struct {
	Total                      int64                     `json:"total" gorm:"column:total"`
	Success                    int64                     `json:"success" gorm:"column:success"`
	Failed                     int64                     `json:"failed" gorm:"column:failed"`
	Failover                   int64                     `json:"failover" gorm:"column:failover"`
	Stream                     int64                     `json:"stream" gorm:"column:stream"`
	AvgAttempts                float64                   `json:"avg_attempts" gorm:"column:avg_attempts"`
	AvgLatency                 float64                   `json:"avg_latency_ms" gorm:"column:avg_latency_ms"`
	InputTokens                int64                     `json:"input_tokens" gorm:"column:input_tokens"`
	OutputTokens               int64                     `json:"output_tokens" gorm:"column:output_tokens"`
	CacheTokens                int64                     `json:"cache_tokens" gorm:"column:cache_tokens"`
	TotalTokens                int64                     `json:"total_tokens" gorm:"column:total_tokens"`
	EstimatedCost              float64                   `json:"estimated_cost" gorm:"column:estimated_cost"`
	FinalSuccessCost           float64                   `json:"final_success_cost" gorm:"column:final_success_cost"`
	TotalAttemptCost           float64                   `json:"total_attempt_cost" gorm:"column:total_attempt_cost"`
	FailedAttemptEstimatedCost float64                   `json:"failed_attempt_estimated_cost" gorm:"column:failed_attempt_estimated_cost"`
	Cost                       float64                   `json:"cost" gorm:"column:cost"`
	StatusBuckets              []RequestTraceAuditBucket `json:"status_buckets" gorm:"-"`
	SourceBuckets              []RequestTraceAuditBucket `json:"source_buckets" gorm:"-"`
	ModelBuckets               []RequestTraceAuditBucket `json:"model_buckets" gorm:"-"`
	ServiceTierBuckets         []RequestTraceAuditBucket `json:"service_tier_buckets" gorm:"-"`
}

type RequestTraceDetail struct {
	Trace    RequestTrace     `json:"trace"`
	Attempts []RequestAttempt `json:"attempts"`
}

File: web/src/api/endpoints/log.ts
import type { InfiniteData } from '@tanstack/react-query';
import { useInfiniteQuery, useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { apiClient, API_BASE_URL } from '../client';
import { logger } from '@/lib/logger';

export type AttemptStatus = 'success' | 'failed' | 'circuit_break' | 'skipped';

export type RelayLogWSMode = 'fresh' | 'continuation' | 'replay';

export type RelayLogWSRecovery = 'reconnect' | 'replay' | 'downgrade';

export interface ChannelAttempt {
    channel_id: number;
    channel_key_id?: number;
    key_id?: number;
    channel_name: string;
    site_id?: number;
    site_account_id?: number;
    account_id?: number;
    base_url?: string;
    model_name: string;
    upstream_model?: string;
    request_protocol?: string;
    upstream_protocol?: string;
    response_protocol?: string;
    attempt_num: number;
    attempt_index?: number;
    status: AttemptStatus;
    duration: number;
    duration_ms?: number;
    ttfb_ms?: number;
    total_ms?: number;
    http_status?: number;
    failure_reason?: string;
    retryable?: boolean;
    input_tokens?: number;
    output_tokens?: number;
    cache_tokens?: number;
    input_cost?: number;
    output_cost?: number;
    estimated_cost?: number;
    cost_incurred?: string;
    cost_source?: string;
    service_tier?: string;
    error_summary?: string;
    created_at?: number;
    sticky?: boolean;
    msg?: string;
}

export interface RelayLog {
    id: number;
    trace_id?: string;
    thread_id?: string;
    client_api_key_id?: number;
    group_id?: number;
    time: number;
    request_model_name: string;
    request_api_key_name?: string;
    request_stream?: boolean;
    request_source?: string;
    client_ip?: string;
    channel: number;
    channel_name: string;
    actual_model_name: string;
    final_status?: string;
    final_channel_id?: number;
    final_site_id?: number;
    final_upstream_model?: string;
    attempts_count?: number;
    total_latency_ms?: number;
    input_tokens: number;
    transport_input_tokens?: number | null;
    bill_input_tokens?: number | null;
    cache_read_tokens?: number | null;
    cache_write_tokens?: number | null;
    cache_tokens?: number;
    output_tokens: number;
    ftut: number;
    use_time: number;
    cost: number;
    estimated_cost?: number;
    final_success_cost?: number;
    total_attempt_cost?: number;
    failed_attempt_estimated_cost?: number;
    service_tier?: string;
    request_content: string;
    response_content: string;
    error: string;
    attempts?: ChannelAttempt[];
    total_attempts?: number;
    used_ws?: boolean;
    ws_mode?: RelayLogWSMode | null;
    ws_recovery?: RelayLogWSRecovery | null;
}

export interface LogListParams {
    page?: number;
    page_size?: number;
    start_time?: number;
    end_time?: number;
    time_range?: string;
    channel_ids?: string;
    model?: string;
    trace_id?: string;
    api_key?: string;
    api_key_id?: number;
    status?: string;
    http_status?: string;
    failure_reason?: string;
    protocol?: string;
    source?: string;
    stream?: boolean;
    failover?: boolean;
    cache_hit?: boolean;
    sort_by?: string;
    sort_order?: string;
}

export type LogListFilters = Omit<LogListParams, 'page' | 'page_size'>;

export interface LogListResponse {
    items: RelayLog[];
    total: number;
    page: number;
    page_size: number;
    has_more: boolean;
}

export interface ActiveRequestSnapshot {
    id: string;
    api_key_id?: number;
    group_id?: number;
    request_model: string;
    request_source?: string;
    request_stream: boolean;
    client_ip?: string;
    started_at: number;
    updated_at: number;
    elapsed_ms: number;
    phase: string;
    channel_id?: number;
    channel_name?: string;
    channel_key_id?: number;
    model_name?: string;
    site_id?: number;
    site_account_id?: number;
    attempts_count?: number;
    last_status?: string;
    last_http_status?: number;
    last_failure_reason?: string;
    written: boolean;
    first_token_seen: boolean;
    used_ws: boolean;
}

export interface ActiveRequestListResponse {
    total: number;
    updated_at: number;
    items: ActiveRequestSnapshot[];
}

export interface ActiveRequestEvent {
    type: 'snapshot' | 'started' | 'updated' | 'completed';
    updated_at: number;
    snapshot?: ActiveRequestSnapshot;
    list?: ActiveRequestListResponse;
}

const activeRequestsQueryKey = ['logs', 'active'] as const;

function mergeActiveRequestEvent(
    current: ActiveRequestListResponse | undefined,
    event: ActiveRequestEvent,
): ActiveRequestListResponse {
    if (event.type === 'snapshot' && event.list) {
        return event.list;
    }

    const updatedAt = event.updated_at || Date.now();
    const base = current ?? { total: 0, updated_at: updatedAt, items: [] };
    const snapshot = event.snapshot;
    if (!snapshot?.id) {
        return { ...base, updated_at: updatedAt };
    }

    const items = event.type === 'completed'
        ? base.items.filter((item) => item.id !== snapshot.id)
        : [snapshot, ...base.items.filter((item) => item.id !== snapshot.id)];

    items.sort((left, right) => {
        if ((left.elapsed_ms ?? 0) === (right.elapsed_ms ?? 0)) {
            return (left.started_at ?? 0) - (right.started_at ?? 0);
        }
        return (right.elapsed_ms ?? 0) - (left.elapsed_ms ?? 0);
    });

    return {
        total: items.length,
        updated_at: updatedAt,
        items,
    };
}

export function useClearLogs() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async () => {
            return apiClient.delete<null>('/api/v1/log/clear');
        },
        onSuccess: () => {
            logger.log('relay logs cleared');
            queryClient.invalidateQueries({ queryKey: ['logs'] });
            queryClient.invalidateQueries({ queryKey: ['request-traces'] });
        },
        onError: (error) => {
            logger.error('failed to clear relay logs:', error);
        },
    });
}

export function useLogDetail(logID: number | undefined, enabled: boolean) {
    return useQuery({
        queryKey: ['logs', 'detail', logID],
        queryFn: async () => {
            if (!logID) throw new Error('missing log id');
            return apiClient.get<RelayLog>(`/api/v1/log/detail/${logID}`);
        },
        enabled: enabled && !!logID,
        staleTime: 60000,
    });
}

export function useActiveRequests(options: { refetchIntervalMs?: number | false; streamEvents?: boolean } = {}) {
    const { refetchIntervalMs = false, streamEvents = false } = options;
    const queryClient = useQueryClient();
    const eventSourceRef = useRef<EventSource | null>(null);
    const [isStreamConnected, setIsStreamConnected] = useState(false);
    const [streamError, setStreamError] = useState<Error | null>(null);

    const query = useQuery({
        queryKey: activeRequestsQueryKey,
        queryFn: async () => apiClient.get<ActiveRequestListResponse>('/api/v1/log/active'),
        refetchInterval: refetchIntervalMs,
        refetchIntervalInBackground: false,
        staleTime: 1000,
    });

    useEffect(() => {
        if (!streamEvents) {
            eventSourceRef.current?.close();
            eventSourceRef.current = null;
            return;
        }

        let cancelled = false;
        let retryTimer: ReturnType<typeof setTimeout> | null = null;
        let retryAttempt = 0;

        const scheduleReconnect = () => {
            if (cancelled) return;
            const delay = Math.min(30000, 1000 * 2 ** retryAttempt);
            retryAttempt += 1;
            retryTimer = setTimeout(() => {
                retryTimer = null;
                connect(true);
            }, delay);
        };

        const connect = async (isReconnect = false) => {
            try {
                const { token } = await apiClient.get<{ token: string }>('/api/v1/log/stream-token');
                if (cancelled) return;

                const eventSource = new EventSource(`${API_BASE_URL}/api/v1/log/active-stream?token=${token}`);
                eventSourceRef.current = eventSource;

                eventSource.onopen = () => {
                    retryAttempt = 0;
                    setIsStreamConnected(true);
                    setStreamError(null);
                    if (isReconnect) {
                        queryClient.invalidateQueries({ queryKey: activeRequestsQueryKey });
                    }
                };

                eventSource.onmessage = (event) => {
                    try {
                        const activeEvent: ActiveRequestEvent = JSON.parse(event.data);
                        queryClient.setQueryData(
                            activeRequestsQueryKey,
                            (old: ActiveRequestListResponse | undefined) => mergeActiveRequestEvent(old, activeEvent),
                        );
                    } catch (e) {
                        logger.error('failed to parse active request stream event:', e);
                    }
                };

                eventSource.onerror = () => {
                    setIsStreamConnected(false);
                    setStreamError(new Error('active request stream disconnected'));
                    eventSource.close();
                    eventSourceRef.current = null;
                    scheduleReconnect();
                };
            } catch (e) {
                if (cancelled) return;
                setStreamError(e instanceof Error ? e : new Error('failed to create active request stream'));
                logger.error('failed to create active request stream:', e);
                scheduleReconnect();
            }
        };

        connect(false);

        return () => {
            cancelled = true;
            if (retryTimer) clearTimeout(retryTimer);
            eventSourceRef.current?.close();
            eventSourceRef.current = null;
            setIsStreamConnected(false);
        };
    }, [queryClient, streamEvents]);

    return {
        ...query,
        isStreamConnected,
        streamError,
    };
}

const logsInfiniteQueryKey = (pageSize: number, filters: LogListFilters) => ['logs', 'infinite', pageSize, filters] as const;

function cleanLogFilters(filters: LogListFilters = {}): LogListFilters {
    const out: Record<string, string | number | boolean> = {};
    for (const [key, value] of Object.entries(filters)) {
        if (value === undefined || value === null || value === '') continue;
        out[key] = value as string | number | boolean;
    }
    return out as LogListFilters;
}

function buildLogParams(page: number, pageSize: number, filters: LogListFilters): Record<string, string | number | boolean> {
    return {
        ...filters,
        page,
        page_size: pageSize,
    };
}

function normalizeLogListResponse(result: RelayLog[] | LogListResponse | null | undefined, page: number, pageSize: number): LogListResponse {
    if (Array.isArray(result)) {
        return {
            items: result,
            total: result.length,
            page,
            page_size: pageSize,
            has_more: result.length >= pageSize,
        };
    }
    if (!result) {
        return { items: [], total: 0, page, page_size: pageSize, has_more: false };
    }
    return {
        items: result.items ?? [],
        total: result.total ?? 0,
        page: result.page ?? page,
        page_size: result.page_size ?? pageSize,
        has_more: Boolean(result.has_more),
    };
}

function hasActiveLogFilters(filters: LogListFilters) {
    return Object.entries(filters).some(([key, value]) => {
        if (value === undefined || value === null || value === '') return false;
        if (key === 'sort_by' && value === 'time') return false;
        if (key === 'sort_order' && value === 'desc') return false;
        return true;
    });
}

export function useLogs(options: {
    pageSize?: number;
    filters?: LogListFilters;
    autoRefresh?: boolean;
    refreshIntervalMs?: number;
    pauseAutoRefresh?: boolean;
} = {}) {
    const {
        pageSize = 20,
        filters = {},
        autoRefresh = false,
        refreshIntervalMs = 10000,
        pauseAutoRefresh = false,
    } = options;

    const cleanFilters = useMemo(() => cleanLogFilters(filters), [filters]);
    const activeFilters = useMemo(() => hasActiveLogFilters(cleanFilters), [cleanFilters]);
    const queryClient = useQueryClient();
    const [isConnected, setIsConnected] = useState(false);
    const [error, setError] = useState<Error | null>(null);
    const eventSourceRef = useRef<EventSource | null>(null);

    const logsQuery = useInfiniteQuery({
        queryKey: logsInfiniteQueryKey(pageSize, cleanFilters),
        initialPageParam: 1,
        queryFn: async ({ pageParam }) => {
            const page = Number(pageParam);
            const result = await apiClient.get<RelayLog[] | LogListResponse | null>(
                '/api/v1/log/list',
                buildLogParams(page, pageSize, cleanFilters),
            );
            return normalizeLogListResponse(result, page, pageSize);
        },
        getNextPageParam: (lastPage, allPages) => {
            if (!lastPage?.has_more) return undefined;
            return allPages.length + 1;
        },
        staleTime: 0,
        refetchOnMount: 'always',
        refetchOnWindowFocus: true,
        refetchInterval: autoRefresh && !pauseAutoRefresh ? refreshIntervalMs : false,
        refetchIntervalInBackground: false,
    });

    const logs = useMemo(() => {
        const pages = logsQuery.data?.pages ?? [];
        const seen = new Set<number>();
        const merged: RelayLog[] = [];

        for (const page of pages) {
            for (const log of page.items) {
                if (seen.has(log.id)) continue;
                seen.add(log.id);
                merged.push(log);
            }
        }

        return merged;
    }, [logsQuery.data]);

    const loadMore = useCallback(async () => {
        if (!logsQuery.hasNextPage || logsQuery.isFetchingNextPage) return;
        try {
            await logsQuery.fetchNextPage();
        } catch (e) {
            logger.error('failed to load more relay logs:', e);
        }
    }, [logsQuery]);

    useEffect(() => {
        if (!autoRefresh || pauseAutoRefresh) {
            eventSourceRef.current?.close();
            eventSourceRef.current = null;
            return;
        }

        let cancelled = false;
        let retryTimer: ReturnType<typeof setTimeout> | null = null;
        let retryAttempt = 0;

        const queryKey = logsInfiniteQueryKey(pageSize, cleanFilters);

        const scheduleReconnect = () => {
            if (cancelled) return;
            const delay = Math.min(30000, 1000 * 2 ** retryAttempt);
            retryAttempt += 1;
            retryTimer = setTimeout(() => {
                retryTimer = null;
                connect(true);
            }, delay);
        };

        const connect = async (isReconnect = false) => {
            try {
                const { token } = await apiClient.get<{ token: string }>('/api/v1/log/stream-token');
                if (cancelled) return;

                const eventSource = new EventSource(`${API_BASE_URL}/api/v1/log/stream?token=${token}`);
                eventSourceRef.current = eventSource;

                eventSource.onopen = () => {
                    retryAttempt = 0;
                    setIsConnected(true);
                    setError(null);
                    if (isReconnect) {
                        queryClient.invalidateQueries({ queryKey });
                    }
                };

                eventSource.onmessage = (event) => {
                    try {
                        const log: RelayLog = JSON.parse(event.data);
                        if (activeFilters) {
                            queryClient.invalidateQueries({ queryKey });
                            return;
                        }
                        queryClient.setQueryData(
                            queryKey,
                            (old: InfiniteData<LogListResponse, number> | undefined) => {
                                if (!old) {
                                    return {
                                        pages: [{
                                            items: [log],
                                            total: 1,
                                            page: 1,
                                            page_size: pageSize,
                                            has_more: false,
                                        }],
                                        pageParams: [1],
                                    };
                                }

                                const exists = old.pages.some((page) => page.items.some((item) => item.id === log.id));
                                if (exists) return old;

                                const first = old.pages[0] ?? {
                                    items: [],
                                    total: 0,
                                    page: 1,
                                    page_size: pageSize,
                                    has_more: false,
                                };
                                const prepended = [log, ...first.items];
                                if (prepended.length > pageSize && old.pages.length > 1) {
                                    queryClient.invalidateQueries({ queryKey });
                                    return {
                                        ...old,
                                        pages: [
                                            {
                                                ...first,
                                                items: prepended.slice(0, pageSize),
                                                total: first.total + 1,
                                                has_more: true,
                                            },
                                            ...old.pages.slice(1),
                                        ],
                                    };
                                }
                                return {
                                    ...old,
                                    pages: [
                                        {
                                            ...first,
                                            items: prepended,
                                            total: first.total + 1,
                                        },
                                        ...old.pages.slice(1),
                                    ],
                                };
                            },
                        );
                    } catch (e) {
                        logger.error('failed to parse relay log stream event:', e);
                    }
                };

                eventSource.onerror = () => {
                    setIsConnected(false);
                    setError(new Error('relay log stream disconnected'));
                    eventSource.close();
                    eventSourceRef.current = null;
                    scheduleReconnect();
                };
            } catch (e) {
                if (cancelled) return;
                setError(e instanceof Error ? e : new Error('failed to create relay log stream'));
                logger.error('failed to create relay log stream:', e);
                scheduleReconnect();
            }
        };

        connect(false);

        return () => {
            cancelled = true;
            if (retryTimer) clearTimeout(retryTimer);
            eventSourceRef.current?.close();
            eventSourceRef.current = null;
            setIsConnected(false);
        };
    }, [activeFilters, autoRefresh, cleanFilters, pageSize, pauseAutoRefresh, queryClient]);

    const clear = useCallback(() => {
        queryClient.removeQueries({ queryKey: logsInfiniteQueryKey(pageSize, cleanFilters) });
    }, [cleanFilters, pageSize, queryClient]);

    return {
        logs,
        total: logsQuery.data?.pages[0]?.total ?? 0,
        isConnected,
        error,
        hasMore: !!logsQuery.hasNextPage,
        isLoading: logsQuery.isLoading,
        isFetching: logsQuery.isFetching,
        isLoadingMore: logsQuery.isFetchingNextPage,
        loadMore,
        refetch: logsQuery.refetch,
        clear,
    };
}
